### DISCUSSION TITLE
compact serialization fails

### DISCUSSION THREAD

**Discussion by marquina91** [panva/jose#50] [Q&A]
Hi,

I'm trying to migrate an application java to nodejs, and for this I decided to use this fantastic library.

but I can't use compact serialization when encrypting it with 2 keys, do you have any idea?

my current code in **java** would be:
  // Initialize the result
    String result = null;

    JsonWebKey jwk = JsonWebKey.Factory.newJwk(new String(Base64.getDecoder().decode(publicKey)));

    // Create a new Json Web Encryption object
    JsonWebEncryption jwe = new JsonWebEncryption();

    // Set the message in the JWE
    jwe.setPlaintext(message);
    // Set the algorithm in the header -> "alg"
    jwe.setAlgorithmHeaderValue(KeyManagementAlgorithmIdentifiers.RSA_OAEP_256);

    // Set the content encryption header -> "enc"
    jwe.setEncryptionMethodHeaderParameter(
        ContentEncryptionAlgorithmIdentifiers.AES_128_CBC_HMAC_SHA_256);
    jwe.setKey(jwk.getKey());

    result = jwe.getCompactSerialization();
    return result;

**and I tried to do it as follows:**

 const recipient = jose.JWK.generateSync('RSA', 2048, {use: 'enc', alg: 'RSA-OAEP-256'});
 
//message: text to encrypt
//publicKey of my application passed to buffer.
 let enc = new jose.JWE.Encrypt(message, {}, {}, publicKey)
 enc.recipient(recipient)
 let result = enc.encrypt('compact'); // this fail

Regards and Thanks in advance.

**Comment by panva** [comment]
Hi @marquina91,

It would generally be immensely helpful if "something fails" report actually came with the error and its trace.

It would save me plenty of time trying to trace your code.

```js
'JWE Compact Serialization doesn\'t support multiple recipients, JWE unprotected headers or AAD'
```

Is this what you're seeing?

Well, no surprise - why are you passing `publicKey` as buffer into the constructor? That argument is [`aad`](https://github.com/panva/jose/blob/master/docs/README.md#new-jweencryptcleartext-protected-unprotected-aad), which is only available with flattened and general json serialization.

If you're only gonna use the compact format use the [`JWE.encrypt()`](https://github.com/panva/jose/blob/master/docs/README.md#jweencryptcleartext-key-protected) shorthand function.

> but I can't use compact serialization when encrypting it with 2 keys, do you have any idea?

Compact cannot be used to encrypt the CEK for multiple recipients. Compact always only has one recipient.

> Only one recipient is supported by the JWE Compact Serialization and it provides no syntax to represent JWE Shared Unprotected Header, JWE Per-Recipient Unprotected Header, or JWE AAD values.

**Comment by marquina91** [comment]
I'm sorry, forgot to put the error

`JWEInvalid: JWE Compact Serialization doesn't support multiple recipients, JWE unprotected headers or AAD`

Just the mistake you mentioned.

I'll take a look at your suggestion, since it looks good.

Thanks for the help and speed.
